extension/property_tree/include/property_tree/data.hpp

```diff
@@ -0,0 +1,201 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_EXT_PROPERTY_TREE_DATA_HPP_
+#define GKO_PUBLIC_EXT_PROPERTY_TREE_DATA_HPP_
+
+
+#include <deque>
+#include <exception>
+#include <list>
+#include <string>
+#include <type_traits>
+
+
+namespace gko {
+namespace extension {
+
+
+class data_s {
+public:
+    enum tag_type { str_t, int_t, double_t, bool_t, empty_t };
+
+    template <tag_type>
+    struct tag {
+        using type = void;
+    };
+
+    /**
+     * Get the data with Type T
+     */
+    template <typename T>
+    T get() const
+    {
+        throw std::runtime_error("Not Supported");
+    }
+
+    template <tag_type T>
+    typename tag<T>::type get() const;
+
+    tag_type get_tag() const { return tag_; }
+
+    /**
+     * Default Constructor
+     */
+    data_s() : tag_(tag_type::empty_t) {}
+
+    /**
+     * Constructor for bool
+     */
+    data_s(bool bb) : tag_(tag_type::bool_t) { bool_ = bb; }
+
+    /**
+     * Constructor for integer
+     */
+    data_s(long long int ii) : tag_(tag_type::int_t) { int_ = ii; }
+
+    /**
+     * Constructor for integer with all integer type
+     */
+    template <typename T, typename = typename std::enable_if<
+                              std::is_integral<T>::value>::type>
+    data_s(T ii) : data_s(static_cast<long long int>(ii))
+    {}
+
+    /**
+     * Constructor for string
+     */
+    data_s(const std::string& str) : tag_(tag_type::str_t) { str_ = str; };
+
+    /**
+     * Constructor for char (otherwise, it will use bool)
+     */
+    data_s(const char* s) : data_s(std::string(s)) {}
+
+    /**
+     * Constructor for double
+     */
+    data_s(double dd) : tag_(tag_type::double_t) { double_ = dd; };
+
+    /**
+     * Constructor for float
+     */
+    data_s(float dd) : data_s(static_cast<double>(dd)) {}
+
+    /**
+     * Overloading stream operator<<
+     */
+    friend std::ostream& operator<<(std::ostream& stream, const data_s& data)
+    {
+        switch (data.tag_) {
+        case tag_type::empty_t:
+            stream << "empty";
+            break;
+        case tag_type::bool_t:
+            stream << (data.bool_ ? "true" : "false");
+            break;
+        case tag_type::int_t:
+            stream << data.int_;
+            break;
+        case tag_type::str_t:
+            stream << data.str_;
+            break;
+        case tag_type::double_t:
+            stream << data.double_;
+            break;
+        }
+        return stream;
+    }
+
+private:
+    tag_type tag_;
+    std::string str_;
+    long long int int_;
+    double double_;
+    bool bool_;
```

Comment:
a tagged union implementation would look something like
```suggestion
    union {
        long long int int_;
        double double_;
        bool bool_;
    }
```
with placement-new and calling the destructor explicitly, you could also move the `str_` member inside the `union`